### PR TITLE
[Offload] Ensure to load images when the device is used

### PR DESCRIPTION
**Pull request by jdoerfert**
When we use the device, e.g., with an API that interacts with it, we need to ensure the image is loaded and the constructors are executed. Two tests are included to verify we 1) load images and run constructors when needed, and 2) we do so lazily only if the device is actually used.

### CODE REVIEW

**Review by jhuber6 — COMMENTED** [review]
I believe device ctor / dtors are handled at image init / deinit time now. We used to init at first kernel / data call. I'm guessing this just expands that for external users?

File: offload/src/PluginManager.cpp [hunk 1]
```diff
@@ -286,13 +291,194 @@ void PluginManager::unregisterLib(__tgt_bin_desc *Desc) {
...
+    return Rc;
+
+  static Int32Envar DumpOffloadEntries =
+      Int32Envar("OMPTARGET_DUMP_OFFLOAD_ENTRIES", -1);
+  if (DumpOffloadEntries.get() == DeviceId)
+    Device.dumpOffloadEntries();
+
+  return OFFLOAD_SUCCESS;
+}
+
```
Comment (jdoerfert) [inline offload/src/PluginManager.cpp:463]: All the code above is just moved.

File: offload/src/interface.cpp [hunk 2]
```diff
@@ -32,6 +35,43 @@
...
+  }
+
+  if (DeviceID == omp_get_initial_device()) {
+    DP("Device is host (%" PRId64 "), returning as if offload is disabled\n",
+       DeviceID);
+    return true;
+  }
+  return false;
+}
+
```
Comment (jdoerfert) [inline offload/src/interface.cpp:74]: This code is just moved and the last part was split off into PluginManager.cpp.

File: offload/src/PluginManager.cpp [hunk 3]
```diff
@@ -286,13 +291,194 @@ void PluginManager::unregisterLib(__tgt_bin_desc *Desc) {
...
+        __tgt_offload_entry DeviceEntry = Entry;
+        if (Entry.size) {
+          if (Device.RTL->get_global(Binary, Entry.size, Entry.name,
+                                     &DeviceEntry.addr) != OFFLOAD_SUCCESS)
+            REPORT("Failed to load symbol %s\n", Entry.name);
+
+          // If unified memory is active, the corresponding global is a device
+          // reference to the host global. We need to initialize the pointer on
+          // the deive to point to the memory on the host.
```
Comment (jhuber6) [inline offload/src/PluginManager.cpp:353]:
```suggestion
          // the device to point to the memory on the host.
```

File: offload/src/PluginManager.cpp [hunk 4]
```diff
@@ -286,13 +291,194 @@ void PluginManager::unregisterLib(__tgt_bin_desc *Desc) {
...
+           DPxPTR(CurrHostEntry->addr), DPxPTR(CurrDeviceEntry->addr),
+           CurrDeviceEntry->size, CurrDeviceEntry->name);
+        HDTTMap->emplace(new HostDataToTargetTy(
+            (uintptr_t)CurrHostEntry->addr /*HstPtrBase*/,
+            (uintptr_t)CurrHostEntry->addr /*HstPtrBegin*/,
+            (uintptr_t)CurrHostEntry->addr + CurrHostEntry->size /*HstPtrEnd*/,
+            (uintptr_t)CurrDeviceEntryAddr /*TgtAllocBegin*/,
+            (uintptr_t)CurrDeviceEntryAddr /*TgtPtrBegin*/,
+            false /*UseHoldRefCount*/, CurrHostEntry->name,
+            true /*IsRefCountINF*/));
```
Comment (jhuber6) [inline offload/src/PluginManager.cpp:443]: Don't we have utils for void ptr arithmetic?

File: offload/src/PluginManager.cpp [hunk 5]
```diff
@@ -286,13 +291,194 @@ void PluginManager::unregisterLib(__tgt_bin_desc *Desc) {
...
-    return createStringError(
-        inconvertibleErrorCode(),
-        "Device number '%i' out of range, only %i devices available", DeviceNo,
-        ExclusiveDevicesAccessor->size());
+  DeviceTy *DevicePtr;
+  {
+    auto ExclusiveDevicesAccessor = getExclusiveDevicesAccessor();
+    if (DeviceNo >= ExclusiveDevicesAccessor->size())
+      return createStringError(
+          inconvertibleErrorCode(),
```
Comment (jhuber6) [inline offload/src/PluginManager.cpp:470]:
```suggestion
```
There's an overload that just takes the string which will add the error code for you. Maybe we should save that for a global cleanup?

File: offload/src/PluginManager.cpp (hunk 4 above)
Comment (jdoerfert) [reply offload/src/PluginManager.cpp:443]: I just moved the code. Cleanup is separate.

File: offload/src/PluginManager.cpp (hunk 5 above)
Comment (jdoerfert) [reply offload/src/PluginManager.cpp:470]: Yes. Let's not conflate this too much. This part of the code is just moved, we should simplify the things afterwards, e.g., pointer arith and clearer APIs.

**Review by jhuber6 — APPROVED** [review]

**Pull request merged**